src/tests.rs

```diff
@@ -1456,6 +1480,491 @@ fn test_secret_redaction() {
     assert_eq!("ClientSecret([redacted])", format!("{:?}", secret));
 }
 
+fn new_device_auth_details(expires_in: u32) -> StandardDeviceAuthorizationResponse {
+    let body = format!(
+        "{{\
+        \"device_code\": \"12345\", \
+        \"verification_uri\": \"https://verify/here\", \
+        \"user_code\": \"abcde\", \
+        \"verification_uri_complete\": \"https://verify/here?abcde\", \
+        \"expires_in\": {}, \
+        \"interval\": 1 \
+        }}",
+        expires_in
+    );
+
+    let device_auth_url =
+        DeviceAuthorizationUrl::new("https://deviceauth/here".to_string()).unwrap();
+
+    let client = new_client().set_device_authorization_url(device_auth_url.clone());
+    client
+        .exchange_device_code()
+        .add_extra_param("foo", "bar")
+        .add_scope(Scope::new("openid".to_string()))
+        .request(mock_http_client(
+            vec![
+                (ACCEPT, "application/json"),
+                (CONTENT_TYPE, "application/x-www-form-urlencoded"),
+                (AUTHORIZATION, "Basic YWFhOmJiYg=="),
+            ],
+            "scope=openid&foo=bar",
+            Some(device_auth_url.url().to_owned()),
+            HttpResponse {
+                status_code: StatusCode::OK,
+                headers: vec![(
+                    CONTENT_TYPE,
+                    HeaderValue::from_str("application/json").unwrap(),
+                )]
+                .into_iter()
+                .collect(),
+                body: body.into_bytes(),
+            },
+        ))
+        .unwrap()
+}
+
+struct IncreasingTime {
+    times: std::ops::RangeFrom<i64>,
+}
+
+impl IncreasingTime {
+    fn new() -> Self {
+        Self { times: (0..) }
+    }
+    fn next(&mut self) -> DateTime<Utc> {
+        let next_value = self.times.next().unwrap();
+        let naive = chrono::NaiveDateTime::from_timestamp(next_value, 0);
+        DateTime::<Utc>::from_utc(naive, chrono::Utc)
+    }
+}
+
+/// Creates a time function that increments by one second each time.
+fn mock_time_fn() -> impl Fn() -> DateTime<Utc> + Send + Sync {
+    let timer = std::sync::Mutex::new(IncreasingTime::new());
+    move || timer.lock().unwrap().next()
+}
+
+/// Mock sleep function that doesn't actually sleep.
+fn mock_sleep_fn(dur: Duration) {
+    println!("Mock sleep for {:?}", dur);
```

Comment:
I'm guessing this was handy for debugging, but it's probably too verbose to print it unconditionally on each test run. I think let's just make the sleep function a no-op (i.e., `|_| ()`)

Reply:
I kept the name `mock_sleep_fn` but removed the println!, as I think it's more maintainable than an unnamed closure. Hope that's ok.